src/Cli/dotnet/NugetPackageDownloader/NuGetPackageDownloader.cs

```diff
@@ -70,18 +87,56 @@ public NuGetPackageDownloader(DirectoryPath packageInstallDir, IFilePermissionSe
                 resolvedPackageVersion,
                 destinationStream,
                 _cacheSettings,
-                _logger,
+                _verboseLogger,
                 cancellationToken);
 
             if (!success)
             {
                 throw new NuGetPackageInstallerException(
-                    $"Downloading {packageId} version {packageVersion.ToNormalizedString()} failed");
+                    string.Format("Downloading {0} version {1} failed", packageId, packageVersion.ToNormalizedString()));
             }
 
+            VerifySigning(nupkgPath);
+
             return nupkgPath;
         }
 
+        private void VerifySigning(string nupkgPath)
+        {
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                if (_firstPartyNuGetPackageSigningVerifier.IsExecutableIsFirstPartySignedWithoutValidation(new FilePath(
+                    typeof(DotNet.Cli.Program).Assembly.Location)))
+                {
+                    if (!_firstPartyNuGetPackageSigningVerifier.Verify(new FilePath(nupkgPath),
+                        out string commandOutput))
+                    {
+                        throw new NuGetPackageInstallerException(LocalizableStrings.FailedToValidatePackageSigning +
```

Comment:
Do we just leave the unsigned package on the machine in this case? Is that a problem, should we delete it/ clean up after ourselves, especially if it's unsigned?

Reply:
good point. We should delete it